### DISCUSSION TITLE
How to use DownloadFileAsync()

### DISCUSSION THREAD

**Discussion by GAT-MAN** [Jericho/ZoomNet#200] [Q&A]
I'm trying to download a video file and keep getting a taskcanceled error, 
I can get the download url filename, so authentication is ok,

```vb.net
    Dim apiKey = "abc" Dim apiSecret = "123" 
    Dim connectionInfo As New JwtConnectionInfo(apiKey, apiSecret)  
    Dim azoomClient = New ZoomClient(connectionInfo)
    Dim v As Object = azoomClient.CloudRecordings.GetRecordingInformationAsync(MeetingID) 
    Dim modelsRecording = Await v
   downloadFileName = CStr(modelsRecording.RecordingFiles.Where(Function(z) 
   z.FileType = Models.RecordingFileType.Video)(0).DownloadUrl)
```

but when i try this 

```vb.net
sourceStream = Await azoomClient.CloudRecordings.DownloadFileAsync(fdownloadFileName, ct).ConfigureAwait(False)
DumpStream(sourceStream, DestFileName)

Private Async Function DumpStream(ByVal outStream As Stream, ByVal outputFileName As String) As Task
    Try

        '  Dump the contents of a stream to a file
        outStream.Flush()
        Dim SavePos As Long = outStream.Position        '  Save the original position in the stream
        outStream.Seek(0, SeekOrigin.Begin)
        Dim f As Stream = File.OpenWrite(outputFileName)
        CopyStream(outStream, f)
        outStream.Position = SavePos                     '  Go back to the original postion in the stream
        f.Close()
    Catch ex As Exception
        MessageBox.Show("Error:DumpStream()>" & ex.Message)
    End Try
End Function

Public Shared Sub CopyStream(ByVal input As Stream, ByVal output As Stream)

    Try
        '  Copy the contents of one stream to another stream
        Dim buf As Byte() = New Byte(8 * 1024 - 1) {}      '  A buffer for storing data while copying
        Dim len As Integer
        len = input.Read(buf, 0, buf.Length)
        While len > 0
            output.Write(buf, 0, len)
            len = input.Read(buf, 0, buf.Length)
        End While
    Catch ex As Exception
        MessageBox.Show("Error:CopyStream()>" & ex.Message)
    End Try
End Sub
```

i get taskcanceled error, do i need to do something with the token for a  callback ?, is there any examples anywhere, i don't have a lot of experience 

Thanks

**Comment by Jericho** [comment]
You can see an example [here](https://github.com/Jericho/ZoomNet/blob/develop/Source/ZoomNet.IntegrationTests/Tests/CloudRecordings.cs#L21-L26) where we retrieve the recording files for a given user. As far as writing the content of a stream, it should be as simple as something like this:

```csharp
using (var outStream = File.OpenWrite("path to new file"))
{
    stream.CopyTo(outStream);
}
```

with my limited knowledge of VB.NET, I was able to convert the C# code to this:
```vb.net
Dim outStream As new FileStream("path to new file", FileMode.Create)
sourceStream.CopyTo(outStream )
```

Let me know if this helps.

**Reply by GAT-MAN** [reply]
Hello Jericho 
Your reply is very appreciated,
i've tried the above code and many other ways, but i can't get past the
`Dim sourceStream = Await azoomClient.CloudRecordings.DownloadFileAsync(fdownloadFileName, ct)`

keeps throwing task cancelled after about 90 seconds 

i got this to work with the download url name

`Dim htc As New WebClient`
`htc.DownloadFileAsync(New Uri(fdownloadFileName), DestFileName)`

??

**Reply by Jericho** [reply]
Maybe the file you are trying to download is extremely large and therefore the download times out after 90 seconds?

**Reply by GAT-MAN** [reply]
The file is 740 mb., is there a setting to extend the timeout.

**Comment by Jericho** [comment]
ZoomNet relies on the default timeout HttpClient settings but you certainly can provide your own HttpClient and configure the timeout to any value you want. 

Something like this:
```csharp
var myHttpClient = new HttpClient()
{
    Timeout = TimeSpan.FromMinutes(10) // This is an exceedingly long timeout for demonstration purposes.
};

var apiKey = Environment.GetEnvironmentVariable("ZOOM_JWT_APIKEY", EnvironmentVariableTarget.User);
var apiSecret = Environment.GetEnvironmentVariable("ZOOM_JWT_APISECRET", EnvironmentVariableTarget.User);
var connectionInfo = new JwtConnectionInfo(apiKey, apiSecret);

var client = new ZoomClient(connectionInfo, myHttpClient);
```

**Reply by GAT-MAN** [reply]
Thanks i'll try that

**Reply by Jericho** [reply]
I reviewed the code in the `DownloadFileAsync` method to make sure there isn't any logic that may contribute (or cause) the problem you are experiencing and I noticed that the code is not optimal. The code currently loads the whole file into memory before returning the stream to your code which is far from optimal when you are dealing with a 740mb file, especially if all you care about is to save the content of this file to disk. In this scenario there is no reason to load the whole file in memory.

Increasing the timeout as you and I have been discussing can probably reduce the likelihood of this problem but it doesn't solve the underlying inefficiency. I have published a [beta package](https://www.myget.org/feed/jericho/package/nuget/ZoomNet/0.42.3-download-files-m0005) with an improved logic that allows downloading files much more efficiently. Let me know if you have an opportunity to try it and if it solves the problem you are experiencing.

**Reply by GAT-MAN** [reply]
I installed the Betapackage 
it did download about half the file but

now it throws this exception n visual studio when i run it

Managed Debugging Assistant 'ContextSwitchDeadlock' : 'The CLR has been unable to transition from COM context 0xb9d600 to COM context 0xb9d548 for 60 seconds. The thread that owns the destination context/apartment is most likely either doing a non pumping wait or processing a very long running operation without pumping Windows messages. This situation generally has a negative performance impact and may even lead to the application becoming non responsive or memory usage accumulating continually over time. To avoid this problem, all single threaded apartment (STA) threads should use pumping wait primitives (such as CoWaitForMultipleHandles) and routinely pump messages during long running operations.'

**Reply by Jericho** [reply]
Maybe this can help: https://stackoverflow.com/questions/578357/visual-studio-contextswitchdeadlock

Also, my guess is that you are running async code in your app's UI thread. Am I right? Is that deliberate and necessary? For example, do you need to update the UI after the async method completes? If the answer is no, you can can probably alleviate this problem by adding `.ConfigureAwait(false)` on the line where you invoke async methods like this example:

```csharp
var stream = await client.CloudRecordings.DownloadFileAsync(recordingFile, cancellationToken).ConfigureAwait(false);
```

**Reply by GAT-MAN** [reply]
Thanks a lot,
I used the  .configurationAwait(False) 
and set the client timeout to above the four minutes it takes to download the file

Much appreciated :)
